### PR TITLE
Leader revocation

### PR DESCRIPTION
**Pull request by jhump**
This stuff is really horrible. No real abstractions with lots of weird sharing of responsibility between `ZkController`, `LeaderElector`, and the actual `ElectionContext` instances :(

For now, I've just done a targeted change so that we can delete nodes in an election and the leader will notice if it loses leadership. I've just added one test but made sure the rest of the suite still passes. I went through the code quite a bit to try to understand the election flow and make sure this was safe.

I was mostly worried about weird edge cases where there could be concurrent calls to `LeaderElector.joinElection()` for the same elector, since I've added some calls to `retryElection` (which will in turn add new call paths from which `joinElection` can be called). But I think I've convinced myself that what's here is safe and correct. The poor abstractions/separation of responsibility makes it hard to safely serialize -- it would require `LeaderElector` and its `ElectionContext` to both share the same mutex (or better yet: re-write many of the flows so simple synchronization and/or thread confinement can be relied on...). So, without a mutex, I'm relying on the fact that it's a state machine, and I haven't added new transitions that will cause confusion/race conditions.

I hope to dig into a more substantial refactor of this ball of wax later.

I've based this pull request on the 5.5.3 head, assuming that we'd wait until we get a 5.5.3 upgrade in prod before trying to push this out.

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/ElectionContext.java [hunk 1]
```diff
@@ -179,6 +179,11 @@ void runLeaderProcess(boolean weAreReplacement, int pauseBeforeStartMs)
     ZkCmdExecutor zcmd = new ZkCmdExecutor(30000);
     zcmd.ensureExists(parent, zkClient);
     final byte[] json = Utils.toJSON(leaderProps);
+    final String path = leaderSeqPath.get();
+    if (path == null) {
+      // concurrently cancelled; just bail
+      return;
+    }
```
Comment (dragonsinth) [inline solr/core/src/java/org/apache/solr/cloud/ElectionContext.java]: should this be down inside of the retry block?

File: solr/core/src/java/org/apache/solr/cloud/ElectionContext.java [hunk 2]
```diff
@@ -63,7 +64,7 @@
   final ZkNodeProps leaderProps;
   final String id;
   final String leaderPath;
-  volatile String leaderSeqPath;
+  final AtomicReference<String> leaderSeqPath = new AtomicReference<>();
```
Comment (dragonsinth) [inline solr/core/src/java/org/apache/solr/cloud/ElectionContext.java]: I found a few instances where the AtomicRef is used (directly) in a log message, but there is a local `path` variable that would have the correct value in it.

File: solr/core/src/java/org/apache/solr/cloud/LeaderElector.java [hunk 3]
```diff
@@ -131,14 +136,34 @@ private void checkIfIamLeader(final ElectionContext context, boolean replacement
     }
 
     if (leaderSeqNodeName.equals(seqs.get(0))) {
-      // I am the leader
+      // I am the leader - watch my node in case leadership is somehow revoked
+      try {
+        ElectionWatcher watcher = new ElectionWatcher(path, path, getSeq(path), context);
+        if (zkClient.exists(path, watcher, true) == null) {
```
Comment (dragonsinth) [inline solr/core/src/java/org/apache/solr/cloud/LeaderElector.java]: There's a super weird case to think through here.  `exists` can be problematic because it really does two different things:
1) If the node exists, leave a watch that triggers on deletion
2) If the node does not exist, leave a watch that triggers on creation

The first feature you want; the second feature you almost certainly don't.  It might be more solid here to to do a getData() call instead and catch NoNodeExeception, so that you only leave a watch on the success case, and don't leave a dangling watch on the missing node case.

File: solr/core/src/java/org/apache/solr/cloud/LeaderElector.java [hunk 4]
```diff
@@ -150,13 +175,14 @@ private void checkIfIamLeader(final ElectionContext context, boolean replacement
       }
       try {
         String watchedNode = holdElectionPath + "/" + toWatch;
-        zkClient.getData(watchedNode, watcher = new ElectionWatcher(context.leaderSeqPath, watchedNode, getSeq(context.leaderSeqPath), context), null, true);
+        ElectionWatcher watcher = new ElectionWatcher(path, watchedNode, getSeq(path), context);
+        if (zkClient.exists(watchedNode, watcher, true) == null) {
+          // the previous node disappeared, check if we are the leader again
+          checkIfIamLeader(context, true);
+        }
```
Comment (dragonsinth) [inline solr/core/src/java/org/apache/solr/cloud/LeaderElector.java]: Same comment here, the previous formulation was likely on purpose.

File: solr/core/src/test/org/apache/solr/cloud/LeaderElectionTest.java [hunk 5]
```diff
@@ -82,17 +83,24 @@ public void setUp() throws Exception {
...
-      super (leaderElector, shardId, collection, coreNodeName, props, zkStateReader);
+      super(leaderElector, shardId, collection, coreNodeName, props, zkStateReader);
+      this.zkStateReader = zkStateReader;
       this.runLeaderDelay = runLeaderDelay;
     }
 
     @Override
+    public TestLeaderElectionContext copy() {
+      return new TestLeaderElectionContext(leaderElector, shardId, collection, id, leaderProps, zkStateReader, runLeaderDelay);
+    }
```
Comment (dragonsinth) [inline solr/core/src/test/org/apache/solr/cloud/LeaderElectionTest.java]: unsupported as per previous superclass behavior?

File: solr/core/src/java/org/apache/solr/cloud/LeaderElector.java (hunk 3 above)
Comment (jhump) [inline solr/core/src/java/org/apache/solr/cloud/LeaderElector.java]: Yeah, that's fair enough about potentially leaking an exists watch (though I think the cost of that is low unless a single session creates a lot of them, and this is actually quite unlikely to happen).

The [recommended recipe](https://zookeeper.apache.org/doc/trunk/recipes.html#sc_recipes_Locks) suggests using an `exists` watch, but I'm guessing that's just to avoid downloading any data in the node. In this case, we're not storing anything in the node so there's nothing really wasteful about the `getData` call. So I'll change the original incantation back to the way it was and update this new one, too.

File: solr/core/src/test/org/apache/solr/cloud/LeaderElectionTest.java (hunk 5 above)
Comment (jhump) [inline solr/core/src/test/org/apache/solr/cloud/LeaderElectionTest.java]: There was nothing relying on the previous superclass behavior. Also, I've now added a test that calls `retryElection`, which requires an actual (non-explosive) implementation of `copy`.

File: solr/core/src/java/org/apache/solr/cloud/ElectionContext.java (hunk 2 above)
Comment (jhump) [inline solr/core/src/java/org/apache/solr/cloud/ElectionContext.java]: Good catch! Fixed. Turns out that the new test had a bad assertion thanks to the same problem -- looking at the `AtomicReference` instead of its string value. Doh! :(

File: solr/core/src/java/org/apache/solr/cloud/ElectionContext.java (hunk 1 above)
Comment (jhump) [inline solr/core/src/java/org/apache/solr/cloud/ElectionContext.java]: Good idea. Moved

File: solr/core/src/java/org/apache/solr/cloud/LeaderElector.java (hunk 3 above)
Comment (jhump) [inline solr/core/src/java/org/apache/solr/cloud/LeaderElector.java]: Fixed

File: solr/core/src/java/org/apache/solr/cloud/LeaderElector.java (hunk 4 above)
Comment (jhump) [inline solr/core/src/java/org/apache/solr/cloud/LeaderElector.java]: changed back

**Comment by jhump** [comment]
Full test suite just passed w/ latest changes. PTAL

**Comment by dragonsinth** [comment]
@jhump did you ever file an issue for this?  I can't find it in the solr tracker

**Comment by jhump** [comment]
@dragonsinth, it's https://issues.apache.org/jira/browse/SOLR-10276

I never submitted this as a patch. I think I actually found a bug in this that I wanted to fix before upstreaming (not a bug that actively causes us any problems -- just a bug that makes the intended fix incomplete).

**Comment by dragonsinth** [comment]
Ah yes... I actually ran into that behavior in production when I was trying to elect solr-1.  If you sort the election nodes into the right order and delete them in order, everything goes fine.  But if you try to "delete all the nodes except my intended leader" it doesn't work right.  I'm not sure exactly what happens, but solr-1 kept deleting its own entry and re-registering later.  I think I just confused everyone.

That said, you should still submit an initial patch to the jira!
